csrc/preseg_passes/remove_bcast_squeeze.cpp

```diff
@@ -0,0 +1,95 @@
+ */
+// clang-format on
+#include <debug.h>
+#include <ir/utils.h>
+#include <options.h>
+#include <preseg_passes/remove_bcast_squeeze.h>
+#include <root_domain_map.h>
+
+namespace nvfuser::preseg_passes {
+
+namespace {
+inline bool isFusionOutput(Fusion* fusion, Val* val) {
+  const auto& outputs = fusion->outputs();
+  return std::find(outputs.begin(), outputs.end(), val) != outputs.end();
+}
+
+// Remove broadcast-squeeze and squeeze-broadcast patterns
+// TODO: still remove when have intermediate ops between broadcast and squeeze
+void removeBcastSqueeze(Fusion* fusion) {
+  // step-1: find and remove broadcast + squeeze pattern
+  // before: Y = broadcast(X); Z = squeeze(Y);  M = someOp(Z)
+  // after : M = someOp(X)
+  // conditions: (1) broadcast and squeeze have the same dim flags
+  //             (2) Y is only consumed by Z (has only one consumer)
+  // special case: if Z is a fusion output, replace the output with X
+  {
+    std::unordered_map<Val*, Val*> replacement_map;
+    for (auto expr : fusion->exprs()) {
+      if (auto sop = dynamic_cast<SqueezeOp*>(expr)) {
+        if (auto bcast = dynamic_cast<BroadcastOp*>(sop->in()->definition())) {
+          if (bcast->getBroadcastDimFlags() == sop->getSqueezeDimFlags() &&
+              bcast->out()->uses().size() == 1) {
+            if (isFusionOutput(fusion, sop->out())) {
+              fusion->replaceOutput(sop->out(), bcast->in());
+            }
+            replacement_map.insert({sop->out(), bcast->in()});
+          }
+        }
+      }
+    }
+    // Replace if there is any match
+    if (!replacement_map.empty()) {
+      ir_utils::replaceValue(fusion, replacement_map);
+    }
+  }
+  // step-2: find and remove squeeze + broadcast pattern
+  // before: Y = squeeze(X); Z = broadcast(Y);  M = someOp(Z)
+  // after : M = someOp(X)
+  // conditions: (1) broadcast and squeeze have the same dim flags
+  //             (2) Y is only consumed by Z (has only one consumer)
+  // special case: if Z is a fusion output, replace the output with X
+  {
+    std::unordered_map<Val*, Val*> replacement_map;
+    for (auto expr : fusion->exprs()) {
+      if (auto bcast = dynamic_cast<BroadcastOp*>(expr)) {
+        if (auto sop = dynamic_cast<SqueezeOp*>(bcast->in()->definition())) {
+          if (bcast->getBroadcastDimFlags() == sop->getSqueezeDimFlags() &&
+              sop->out()->uses().size() == 1) {
+            if (isFusionOutput(fusion, bcast->out())) {
+              fusion->replaceOutput(bcast->out(), sop->in());
+            }
+            replacement_map.insert({bcast->out(), sop->in()});
+          }
+        }
+      }
+    }
+    // Replace if there is any match
+    if (!replacement_map.empty()) {
+      ir_utils::replaceValue(fusion, replacement_map);
+    }
+  }
+}
+
+} // namespace
+
+void RemoveBcastSqueeze::runPass(Fusion* fusion) {
+  if (isDebugDumpEnabled(DebugDumpOption::PreSegmenterLogging)) {
+    debug() << "Fusion before RemoveBcastSqueeze:" << std::endl;
```

Comment:
Not for this PR, but what do you think about letting https://github.com/NVIDIA/Fuser/blob/1e48b3752f8cfc976c83df5f31b6ddc2b12f6fc6/csrc/preseg_passes/pre_segmenter.cpp#L21 to print fusions after each pass? This way, we don't need to duplicate the printing in each individual pass.

Reply:
Good point. we can do a small refactor after this PR.